<html>
<head>
<title>Error Handling in R</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1 class="translated">R中的错误处理</h1>
<blockquote>原文：<a href="https://medium.com/analytics-vidhya/error-handling-in-r-23b22d82fa6f?source=collection_archive---------21-----------------------#2020-05-07">https://medium.com/analytics-vidhya/error-handling-in-r-23b22d82fa6f?source=collection_archive---------21-----------------------#2020-05-07</a></blockquote><div><div class="ds gw gx gy gz ha"/><div class="hb hc hd he hf"><div class=""/><figure class="ev ex ig ih ii ij er es paragraph-image"><div class="er es if"><img src="../Images/1e3afe527663fceeb8cc1de28b100a2f.png" data-original-src="https://miro.medium.com/v2/resize:fit:800/format:webp/0*LjufwyJLt65_OtrK.jpg"/></div></figure><h1 id="8a12" class="im in hi bd io ip iq ir is it iu iv iw ix iy iz ja jb jc jd je jf jg jh ji jj bi translated">什么是例外？</h1><p id="465d" class="pw-post-body-paragraph jk jl hi jm b jn jo jp jq jr js jt ju jv jw jx jy jz ka kb kc kd ke kf kg kh hb bi translated">当你的代码被执行时，一个不希望出现的情况被称为异常，例如当你的代码试图将一个值除以零时。</p><h1 id="b40d" class="im in hi bd io ip iq ir is it iu iv iw ix iy iz ja jb jc jd je jf jg jh ji jj bi translated">异常处理</h1><p id="b0e2" class="pw-post-body-paragraph jk jl hi jm b jn jo jp jq jr js jt ju jv jw jx jy jz ka kb kc kd ke kf kg kh hb bi translated">异常处理是处理代码中可能出现的错误并避免代码突然停止的过程。用简单的英语来说，我们的代码应该要么通过执行预期的任务来结束，要么在无法完成任务时打印一条有用的消息。</p><p id="569f" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">我们有这个代码，它在列表中有一个非数字值，我们试图用向量v的每个元素除5</p><p id="bfa2" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">#list用一个非数值v </p><figure class="ko kp kq kr fd ij er es paragraph-image"><div class="er es kn"><img src="../Images/cc09d4ffa989b2fd2e091c6743d4ef51.png" data-original-src="https://miro.medium.com/v2/resize:fit:1280/format:webp/0*WmbvofChnSNvQ45n.png"/></div></figure><p id="e4a4" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">Here we can see that the code has not printed any result and has stopped abruptly.</p><p id="cf6b" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">To avoid these situations we use exception handling constructs available in R</p><h1 id="37b1" class="im in hi bd io ip iq ir is it iu iv iw ix iy iz ja jb jc jd je jf jg jh ji jj bi translated">Exception Handling Constructs in R</h1><ol class=""><li id="4340" class="ks kt hi jm b jn jo jr js jv ku jz kv kd kw kh kx ky kz la bi translated">try</li><li id="7c99" class="ks kt hi jm b jn lb jr lc jv ld jz le kd lf kh kx ky kz la bi translated">tryCatch</li></ol><h1 id="bdb7" class="im in hi bd io ip iq ir is it iu iv iw ix iy iz ja jb jc jd je jf jg jh ji jj bi translated">Using try</h1><p id="da13" class="pw-post-body-paragraph jk jl hi jm b jn jo jp jq jr js jt ju jv jw jx jy jz ka kb kc kd ke kf kg kh hb bi translated">We need to enclose the objectionable statements in the try block. The statements passed inside are like arguments to a function. In case you have more than one statements, it is preferred that you write a function with all those statements and call the function inside the try block</p><pre class="ko kp kq kr fd lg lh li lj aw lk bi"><span id="8dec" class="ll in hi lh b fi lm ln l lo lp">v&lt;-list(1,2,4,'0',5)<br/>for (i in v) {<br/>  try(print(5/i))<br/>}</span></pre><figure class="ko kp kq kr fd ij er es paragraph-image"><div class="er es kn"><img src="../Images/28cca558b27d4f8c61ddce38f8038281.png" data-original-src="https://miro.medium.com/v2/resize:fit:1280/format:webp/0*UXQre0jvLJZQC9IE.png"/></div></figure><p id="7c04" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">Using the try block we can see the code ran for all the other cases even after the error in one of the iteration.</p><h1 id="77b7" class="im in hi bd io ip iq ir is it iu iv iw ix iy iz ja jb jc jd je jf jg jh ji jj bi translated">Using tryCatch</h1><p id="9145" class="pw-post-body-paragraph jk jl hi jm b jn jo jp jq jr js jt ju jv jw jx jy jz ka kb kc kd ke kf kg kh hb bi translated">The try block will not let your code stop but does not provide any mechanism to handle the exception. By handling I mean the actions we want to perform if some error occurs in code execution.</p><p id="de69" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">In this case, we can use the <strong class="jm hj"> tryCatch </strong>块来处理异常。</p><p id="60df" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">我们将在<strong class="jm hj"> tryCatch </strong>块中包含不良语句，并向tryCatch传递另一个参数<strong class="jm hj"> error </strong>。</p><p id="1511" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">该错误将一个函数或一个指令作为输入，并且我们可以执行在该函数中出现错误的情况下要执行的所有补救步骤。</p><pre class="ko kp kq kr fd lg lh li lj aw lk bi"><span id="13d2" class="ll in hi lh b fi lm ln l lo lp">v&lt;-list(1,2,4,'0',5)<br/>for (i in v) {<br/>  tryCatch(print(5/i),error=function(e){<br/>    print("Non conformabale arguments")<br/>  })<br/>}</span></pre><figure class="ko kp kq kr fd ij er es paragraph-image"><div class="er es kn"><img src="../Images/48edf34bbc8c5945f018aef2f8f52dca.png" data-original-src="https://miro.medium.com/v2/resize:fit:1280/format:webp/0*2Ks9JmWMXgNNbjfO.png"/></div></figure><p id="d3c7" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">我希望这能帮助你理解r中的错误处理概念。</p><p id="b9b6" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated"><em class="lq">你可能还想阅读:</em></p><ul class=""><li id="b646" class="ks kt hi jm b jn ki jr kj jv lr jz ls kd lt kh lu ky kz la bi translated"><a class="ae lv" href="https://techscouter.blogspot.com/2019/12/looping-in-r.html" rel="noopener ugc nofollow" target="_blank">R中的循环</a></li><li id="c493" class="ks kt hi jm b jn lb jr lc jv ld jz le kd lf kh lu ky kz la bi translated"><a class="ae lv" href="https://techscouter.blogspot.com/2020/03/writing-first-go-program.html" rel="noopener ugc nofollow" target="_blank">编写你的第一个围棋程序</a></li></ul><p id="45ab" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">订阅<a class="ae lv" href="https://techscouter.blogspot.com/" rel="noopener ugc nofollow" target="_blank"> Techscouter </a>获取有关ML、数据挖掘、编程语言和当前市场趋势的文章更新。</p><p id="1c75" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">作者:<a class="ae lv" href="https://www.linkedin.com/in/japneet-singh-chawla/" rel="noopener ugc nofollow" target="_blank">贾普内特·辛格</a></p><p id="9457" class="pw-post-body-paragraph jk jl hi jm b jn ki jp jq jr kj jt ju jv kk jx jy jz kl kb kc kd km kf kg kh hb bi translated">快乐学习</p></div></div>    
</body>
</html>